### PR TITLE
New script burgle.lic

### PR DESCRIPTION
**Pull request by vtcifer**
Support for Breaking and Entering, released to live today.

### CODE REVIEW

File: bne.lic [hunk 1]
```diff
@@ -0,0 +1,76 @@
+=begin
+
+Very simple BnE script.  Use at your own risk - there are VERY high fines if you get caught
```
Comment (rpherbig) [inline bne.lic]: 💡 I have a minor preference for this documentation going on elanthipedia

File: bne.lic [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+class BNE
+  include DRC
+  include DRCT
+  include DRCI
+
+  def initialize
+    start_script('jail-buddy') unless Script.running?('jail-buddy')
+    @settings = get_settings
+    @bne_settings = @settings.bne_settings
```
Comment (rpherbig) [inline bne.lic]: ❌ `bne_settings` needs to be in `base.yaml` and `base-empty.yaml`

File: bne.lic [hunk 3]
```diff
@@ -0,0 +1,76 @@
+=begin
+
+Very simple BnE script.  Use at your own risk - there are VERY high fines if you get caught
+
+#yaml settings examples:
+#bne_settings:
+#  room: nnnn            # room id that bne should be done in.  Guards should never come to this room, so pick carefully
+#  entry_type: loxkpick  # entry method - rope or lockpick.  must have a heavy rope if using rope.  if using lockpick, it will use use_lockpick_ring/lockpick_container settings
+#  search_count: 0       # futue develpment, not supported right now
+#  dispose: false        # futue develpment, not supported right now , when supproted will use stealing_bag for goods
```
Comment (rpherbig) [inline bne.lic]: 💡 'supproted' -> 'supported'

File: bne.lic [hunk 4]
```diff
@@ -0,0 +1,76 @@
+=begin
+
+Very simple BnE script.  Use at your own risk - there are VERY high fines if you get caught
+
+#yaml settings examples:
+#bne_settings:
+#  room: nnnn            # room id that bne should be done in.  Guards should never come to this room, so pick carefully
+#  entry_type: loxkpick  # entry method - rope or lockpick.  must have a heavy rope if using rope.  if using lockpick, it will use use_lockpick_ring/lockpick_container settings
```
Comment (rpherbig) [inline bne.lic]: 💡 'loxkpick' -> 'lockpick'

File: bne.lic [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      elsif bput('get my lockpick', 'You get', 'You are already holding' ,'What were you') =~ /What were you/
+        message("Couldn't find entry item: lockpick")
+        return
+      end
+    else
+      message("Unknown entry method: #{@bne_settings['entry_type']}")
+      return
+    end
+    #setup your buffs for bne
+    wait_for_script_to_complete('buff', ['bne']) if @settings.waggle_sets['bne']
```
Comment (rpherbig) [inline bne.lic]: ❔ Might be worth mentioning in the documentation that it supports a buff set?

File: bne.lic [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        hide_attemps -= 1
+        if hide_attemps <= 0
+          message("Couldn't hide.  Find a better room.")
+          return
+        end
+      end
+    end
+    bput('burgle', 'You make short work of the lock', 'You scale up the side of a wall')
+      #for each room_list  - tbd
+    fput('go window')
```
Comment (rpherbig) [inline bne.lic]: ❔ Can this be a `bput`? Don't worry if it can't

**Comment by rpherbig** [comment]
A bunch of nice-to-have comments. The only thing stopping this from being merged are the settings in base and base-empty.

**Comment by vtcifer** [comment]
All reviewed comments should be applied.

**Comment by dr-leustyin** [comment]
Any chance we can change the name to ;burgle ?

;bne isn't a very intuitive name

File: bne.lic [hunk 7]
```diff
@@ -67,10 +85,67 @@ class BNE
...
+    end
+
+    bput('go window', 'You take a moment to reflect on the caper')
+
+  end
+
+  def search_for_loot(target)
+    if /It looks valuable/ =~ bput("search #{target}",'It looks valuable','Roundtime','I could not')
+      if @bne_settings['dispose']
+        drop_item("#{DRC.right_hand}")
```
Comment (rpherbig) [inline bne.lic]: 💡 The string interpolation isn't necessary here (or line 109)

File: bne.lic [hunk 8]
```diff
@@ -67,10 +85,67 @@ class BNE
...
+
+  def put_item?(item)
+    case bput("put my #{item} in my #{@settings.stealing_bag}", 'What were you', 'You put', "You can't do that", 'no matter how you arrange it', 'The .* is *.* too \w+ to fit in', 'There isn\'t any more room', 'perhaps try doing that again')
+    when 'perhaps try doing that again'
+      return put_item?(item)
+    when 'You put'
+      return true
+    when 'What were you'
+      handheld = held_item(item)
+      drop_item(held_item(item)) if handheld
```
Comment (rpherbig) [inline bne.lic]: 💡 Could this be `drop_item(handheld) if handheld`?

File: bne.lic [hunk 9]
```diff
@@ -67,10 +85,67 @@ class BNE
...
+    if /It looks valuable/ =~ bput("search #{target}",'It looks valuable','Roundtime','I could not')
+      if @bne_settings['dispose']
+        drop_item("#{DRC.right_hand}")
+      else
+        put_item?("#{DRC.right_hand}")
+      end
+    end
+  end
+
+  def put_item?(item)
```
Comment (rpherbig) [inline bne.lic]: 💡 Why does `put_item?` return a boolean if its return value isn't used?

**Comment by rpherbig** [comment]
@vtcifer looks good - I'm ready to merge it. How do you feel about the comment from @dr-leustyin ?

**Comment by Hiinky** [comment]
Might be an edit for later as its more complicated due to the random exits but if we could add sneaking from room to room before searching and jumping out the window, it doubles your stealth mindstates.

**Comment by Dartellum** [comment]
Could you add a section to pause scripts, such as sanowret-crystal or performance? In def initialize: 
    pause_script('performance') if Script.running?('performance')
    pause_script('sanowret-crystal') if Script.running?('sanowret-crystal')

then:
before_dying do
  unpause_script('performance') if Script.running?('performance')
  unpause_script('sanowret-crystal') if Script.running?('sanowret-crystal')
end

etc.

**Comment by BinuDR** [comment]
> 
> 
> Could you add a section to pause scripts, such as sanowret-crystal or performance? In def initialize:
> pause_script('performance') if Script.running?('performance')
> pause_script('sanowret-crystal') if Script.running?('sanowret-crystal')
> 
> then:
> before_dying do
> unpause_script('performance') if Script.running?('performance')
> unpause_script('sanowret-crystal') if Script.running?('sanowret-crystal')
> end
> 
> etc.

It might be a good idea to pause all scripts.  Any accidental interference could be highly problematic.

**Comment by Dartellum** [comment]
> > Could you add a section to pause scripts, such as sanowret-crystal or performance? In def initialize:
> > pause_script('performance') if Script.running?('performance')
> > pause_script('sanowret-crystal') if Script.running?('sanowret-crystal')
> > then:
> > before_dying do
> > unpause_script('performance') if Script.running?('performance')
> > unpause_script('sanowret-crystal') if Script.running?('sanowret-crystal')
> > end
> > etc.
> 
> It might be a good idea to pause all scripts. Any accidental interference could be highly problematic.

put **pause_all** under the start jail-buddy line. :)

**Comment by BinuDR** [comment]
and **unpause_all** in the **before_dying do** section.

**Comment by Dartellum** [comment]
> and **unpause_all** in the **before_dying do** section.

That did not work for me. 
This did though: **Script.running.find_all { |s| s.paused? && !s.no_pause_all }.each(&:unpause)**

**Comment by vtcifer** [comment]
> Any chance we can change the name to ;burgle ?
> 
> ;bne isn't a very intuitive name

IMO, BNE is a pretty common shorthand for breaking and entering (which is the system name technically, and the shorthand most are using already).  I'm a big fan of script names as short as possible. 

> Might be an edit for later as its more complicated due to the random exits but if we could add sneaking from room to room before searching and jumping out the window, it doubles your stealth mindstates.

Once more room searches are put in, I don't think that'll be necessary.  You get additional stealth exp for stealing too, get 1 or 2 items, and it's damn near 34/34 to begin with.  Sneaking uses RT which adds too much danger I think.  I'm open to it though, if it's really needed after searching more than 1 room is in.

> > Could you add a section to pause scripts, such as sanowret-crystal or performance? In def initialize:
> > pause_script('performance') if Script.running?('performance')
> > pause_script('sanowret-crystal') if Script.running?('sanowret-crystal')
> > then:
> > before_dying do
> > unpause_script('performance') if Script.running?('performance')
> > unpause_script('sanowret-crystal') if Script.running?('sanowret-crystal')
> > end
> > etc.
> 
> It might be a good idea to pause all scripts. Any accidental interference could be highly problematic.

Not a bad idea.  does pause-all impact jail-buddy?  Because that should remain running.

File: bne.lic (hunk 9 above)
Comment (vtcifer) [reply bne.lic]: Ripped it out of steal, for future use with bining/pawning.

**Comment by Dartellum** [comment]
> > Any chance we can change the name to ;burgle ?
> > ;bne isn't a very intuitive name
> 
> IMO, BNE is a pretty common shorthand for breaking and entering (which is the system name technically, and the shorthand most are using already). I'm a big fan of script names as short as possible.
> 
> > Might be an edit for later as its more complicated due to the random exits but if we could add sneaking from room to room before searching and jumping out the window, it doubles your stealth mindstates.
> 
> Once more room searches are put in, I don't think that'll be necessary. You get additional stealth exp for stealing too, get 1 or 2 items, and it's damn near 34/34 to begin with. Sneaking uses RT which adds too much danger I think. I'm open to it though, if it's really needed after searching more than 1 room is in.
> 
> > > Could you add a section to pause scripts, such as sanowret-crystal or performance? In def initialize:
> > > pause_script('performance') if Script.running?('performance')
> > > pause_script('sanowret-crystal') if Script.running?('sanowret-crystal')
> > > then:
> > > before_dying do
> > > unpause_script('performance') if Script.running?('performance')
> > > unpause_script('sanowret-crystal') if Script.running?('sanowret-crystal')
> > > end
> > > etc.
> > 
> > 
> > It might be a good idea to pause all scripts. Any accidental interference could be highly problematic.
> 
> Not a bad idea. does pause-all impact jail-buddy? Because that should remain running.

Jail-buddy stays running. 

I had to use **Script.running.find_all { |s| s.paused? && !s.no_pause_all }.each(&:unpause)** as unpause_all did not work. Trying to figure out why.

File: bne.lic (hunk 7 above)
Comment (vtcifer) [reply bne.lic]: I think I just left that in from when I was using bput and didn't fully clear it when I ripped the code out of steal.  I can fix that with the pausing other scripts too.

**Comment by vtcifer** [comment]
> Jail-buddy stays running.
> 
> I had to use **Script.running.find_all { |s| s.paused? && !s.no_pause_all }.each(&:unpause)** as unpause_all did not work. Trying to figure out why.

Actually, may be better to iterate and store all unpaused scripts to pause anyway, that way we don't unpause something paused for other reasons than this script.

**Comment by Hiinky** [comment]
I vote name burgle.  I haven't used BNE anywhere but since its going to be mostly called by t2 its not terrible.

The only way I've been able to get 17/34 stealth as a survival second or tert is to sneak from room to room and search twice.  Otherwise its only 9/34 and I've not had any trouble getting out.

I would prefer jail-buddy to not work. It defaults to plead innocent while pleading guilty significantly lowers the fine if caught.

**Comment by vtcifer** [comment]
> The only way I've been able to get 17/34 stealth as a survival second or tert is to sneak from room to room and search twice. Otherwise its only 9/34 and I've not had any trouble getting out.

I'll do some more testing with this.  I suspect it'll be quicker/safe to move->hide->search than sneak->search.

> I would prefer jail-buddy to not work. It defaults to plead innocent while pleading guilty significantly lowers the fine if caught.

I think jail-buddy would need to be enhanced in that case.  Due to the fines for getting caught ideally it should never be needed (we want to avoid getting caught 100% of the time), but given the risk of item loss I can't say leaving it out is a good idea.

**Comment by vtcifer** [comment]
I think I've got a good idea for the movement, and want to put it into psuedocode before working on it.  A recursive function I think works best here.  


This goes right after the burgle, before the go window:
```
	if search_max > 0
		@search_counter = 0
		rob_the_place
	end

def rob_the_place(direction=nil)
	visited.add(direction) unless direction = nil
	loot(roomobj) if !blacklist.contains(roomname)
	if search_counter >= search_max return
	while directions.size > visited.size and search_counter < search_max
		pick random direction you haven't been -> newdir
		visited.add(newdir)
		reverse = reverse[newdir]
		move(newdir)
		rob_the_place(reverse)
		move(reverse)
	loop
end

def Move(direction)
	if hidden && search_counter < search_max
		sneak direction
	else
		direction
	end
end

def loot(obj)
	hide if !hidden && !invis
	search obj
	if succesful
		store hands if full
		return
	elsif not succesful && retry=true
		loot(obj)
	end
end
```

One question I have with lich/ruby.  Is there a way with flags to change a variable value when the string occurs (or something similar).

**Comment by BinuDR** [comment]
Here's the list of rooms and their respective surfaces for searching.

https://elanthipedia.play.net/Breaking_and_Entering#Rooms

**Comment by vtcifer** [comment]
I'm not against changing the name from bne to something else, but I'd like avoid burgle, as someone running it manually, and mistyping could result in doing it before the cooldown is up.

**Comment by rpherbig** [comment]
> I'm not against changing the name from bne to something else, but I'd like avoid burgle, as someone running it manually, and mistyping could result in doing it before the cooldown is up.

We've solved this in two ways in other scripts:
1. Require an argument (e.g. `start`) or the script aborts, to prevent accidental executions
1. Check cooldown when the script runs and refuse to run unless an `override` arg is provided

**Comment by rpherbig** [comment]
@vtcifer there has been a lot of great discussion in here. That said, I want to try to differentiate changes necessary to make this script usable, vs those to make it better. The former is all that is needed to merge this. Let me know when you feel like it's in a good place - you can always make it better in future PRs.

**Comment by vtcifer** [comment]
> > I'm not against changing the name from bne to something else, but I'd like avoid burgle, as someone running it manually, and mistyping could result in doing it before the cooldown is up.
> 
> We've solved this in two ways in other scripts:
> 
> 1. Require an argument (e.g. `start`) or the script aborts, to prevent accidental executions
> 2. Check cooldown when the script runs and refuse to run unless an `override` arg is provided

My concern was if they typo without the script start start character. ex, instead of ;burgle, they typo burgle.  Given that in game burgle now defaults to burgle help if you type anything after that isn't help/recall, I think requiring a "Start" arg can put in.
#2 is already taken care of in the script.

> @vtcifer there has been a lot of great discussion in here. That said, I want to try to differentiate changes necessary to make this script usable, vs those to make it better. The former is all that is needed to merge this. Let me know when you feel like it's in a good place - you can always make it better in future PRs.

Excellent point.  I think the last bits before it's MVP, is the script pausing and the typo protection.  Next commit will have those.

Room Movement, blacklisting and bin/pawn support can be in subsequent PRs

**Comment by vtcifer** [comment]
I want to call out for transparency, there is also an update to common.lic here to support ending khri silence with release_invisibility.

**Comment by DRWafflesworth** [comment]
Could this be called with an arg to choose on the spot whether to use a rope or lockpick?

File: burgle.lic [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    ]
+
+    args = parse_args(arg_definitions)
+
+    start_script('jail-buddy') unless Script.running?('jail-buddy')
+    @settings = get_settings
+    @burgle_settings = @settings.burgle_settings
+
+    if !valid_burgle_settings?
+      message("Invalid burgle_settings.  Chcek the documenatation before running, this is a dangerous script to run if it's not understood.")
```
Comment (BinuDR) [inline burgle.lic]: 💡  Chcek -> Check

**Comment by rpherbig** [comment]
✔️ Looks good to me. @vtcifer are you ready for me to merge it?

**Comment by vtcifer** [comment]
> ✔️ Looks good to me. @vtcifer are you ready for me to merge it?

GTG.  I'll update the message typos in the next PR,

**Comment by Hiinky** [comment]
For sure something for a later version but I second the request to be able to send rope or lockpick as an arg if we want to rotate methods on the fly.

**Comment by Dartellum** [comment]
You could change requiring start to requiring lockpick or rope.

**Comment by Stormsocks** [comment]
I'm going to toss in a third for the entry method changing - something like necromancers use for their rituals would be incredible:

cycle, rope, or lockpick.   cycle would choose whichever skill has the lowest exp at the moment.

When additional rooms get added, it would be great if there was a way to set it to loot the special room before the kitchen/bedroom.

**Pull request merged**